src/agent/state-inspector.ts

```diff
@@ -0,0 +1,549 @@
+
+// TODO: Unify some common code with state.ts in future PRs.
+
+import * as inspector from 'inspector';
+import * as lodash from 'lodash';
+import * as util from 'util';
+
+import {debugAssert} from './debug-assert';
+
+const isEmpty = lodash.isEmpty;
+
+import {StatusMessage} from '../status-message';
+
+import * as apiTypes from '../types/api-types';
+import {DebugAgentConfig} from './config';
+import {V8Inspector} from './v8inspector';
+
+const assert = debugAssert(!!process.env.CLOUD_DEBUG_ASSERTIONS);
+
+// Error message indices into the resolved variable table.
+const BUFFER_FULL_MESSAGE_INDEX = 0;
+const NATIVE_PROPERTY_MESSAGE_INDEX = 1;
+const GETTER_MESSAGE_INDEX = 2;
+const ARG_LOCAL_LIMIT_MESSAGE_INDEX = 3;
+
+
+/**
+ * Checks that the provided expressions will not have side effects and
+ * then evaluates the expression in the current execution context.
+ *
+ * @return an object with error and mirror fields.
+ */
+export function evaluate(
+    expression: string, frame: inspector.Debugger.CallFrame,
+    v8inspector: V8Inspector):
+    {error: string|null, object?: inspector.Runtime.RemoteObject} {
+  // First validate the expression to make sure it doesn't mutate state
+  const acorn = require('acorn');
+  try {
+    const ast = acorn.parse(expression, {sourceType: 'script'});
+    const validator = require('./validator');
+    if (!validator.isValid(ast)) {
+      return {error: 'expression not allowed'};
+    }
+  } catch (err) {
+    return {error: err.message};
+  }
+
+  // Now actually ask V8 Inspector to evaluate the expression
+  const result = v8inspector.evaluateOnCallFrame(frame.callFrameId, expression);
+  if (result.error || !result.response) {
+    return {
+      error: result.error ? String(result.error) : 'no reponse in result'
+    };
+  } else if (result.response.exceptionDetails) {
+    return {error: String(result.response.exceptionDetails)};
+  } else {
+    return {error: null, object: result.response.result};
+  }
+}
+
+class StateResolver {
+  private callFrames_: Array<inspector.Debugger.CallFrame>;
+  private v8Inspector_: V8Inspector;
+  private expressions_: string[];
+  private config_: DebugAgentConfig;
+  private scriptmapper_: {[id: string]: any};
+  private breakpoint_: apiTypes.Breakpoint;
+  private evaluatedExpressions_: apiTypes.Variable[];
+  private totalSize_: number;
+  private messageTable_: apiTypes.Variable[];
+  private resolvedVariableTable_: apiTypes.Variable[];
+  private rawVariableTable_: Array<any>;
+
+  /**
+   * @param {Array<!Object>} callFrames
+   * @param {Array<string>} expressions
+   * @param {!Object} config
+   * @constructor
+   */
+  constructor(
+      callFrames: Array<inspector.Debugger.CallFrame>,
+      breakpoint: apiTypes.Breakpoint, config: DebugAgentConfig,
+      scriptmapper: {[id: string]: any}, v8Inspector: V8Inspector) {
+    this.callFrames_ = callFrames;
+    this.breakpoint_ = breakpoint;
+    this.expressions_ = breakpoint.expressions as string[];
```

Comment:
Can you add a TODO to investigate whether this cast can be avoided?

Reply:
Removed the casting now.